### DISCUSSION TITLE
Question: Intended use with venvs

### DISCUSSION THREAD

**Discussion by laenan8466** [pawamoy/pdm-multirun#10] [Q&A] [answered]
Hi,
this are questions concerning the `PDM_MULTIRUN_USE_VENVS` environment variable. 

1) What is the default setting here?
    Guess this should be `0`. Correct?

2) Usage with pyenv/venvs
  **Situation**:
  Normally I would go with one venv, building on one or multiple python versions (build with pyenv).
  I don't like my pyenvs to be bloated by packages from projects - Therefore I use venvs (saved in package folder or App folder). How would I have to setup pdm-multirun to follow that direction?

3) Is this good practice/what is your practice here?
   In [copier-pdm setup script](https://github.com/pawamoy/copier-pdm/blob/main/project/scripts/setup.sh) you are using `__pypackages__` by default, but the [PEP582](https://peps.python.org/pep-0582/) was rejected. How is this intended?

4) On further notice to  [copier-pdm setup script](https://github.com/pawamoy/copier-pdm/blob/main/project/scripts/setup.sh)
    When using in a project:
    ```bash
    export PDM_MULTIRUN_USE_VENVS=1
    export PDM_MULTIRUN_VERSIONS="3.10 3.11 3.12"
    pdm multirun -v pdm install -G:all
    ```
    
    The `3.10` and `3.12` versions/interpreters are skipped. (having set `3.11` to `global` with `pyenv`)
    I would have expected to have created three venvs for different pyenv versions.
    
    When creating the venvs first and setting pdm config `venv.in_project = false` it is working that the three venvs are filled.
    Not working when set `true`, but this is likely an [issue with pdm](https://github.com/pdm-project/pdm/issues/2395).

**Comment by pawamoy** [comment, marked as answer]
(just enabled discussions, and transferred the issue :slightly_smiling_face:) 

> What is the default setting here?

The default setting is *unset*. The option is activated **only** if the value is 1. Any other value (0, empty, 2), and the venv feature will be *disabled*.

> I don't like my pyenvs to be bloated by packages from projects

Definitely something to avoid indeed.

> How would I have to setup pdm-multirun to follow that direction?

The idea is to declare and create a few venvs with PDM, then use the same names with PDM Multirun:

```bash
# Unfortunately I don't think PDM supports creating multiple venv in the project folder,
# so we have to tell it to create them elsewhere. But maybe I'm wrong.
pdm config venv.in_project false

pdm venv create -n 38 python3.8
pdm venv create -n 39 python3.9
pdm venv create -n 310 python3.10
pdm venv create -n 311 python3.11
pdm venv create -n 312 python3.12

export PDM_MULTIRUN_VERSIONS="38 39 310 311 312"
export PDM_MULTIRUN_USE_VENVS=1
```

> Is this good practice/what is your practice here?
In [copier-pdm setup script](https://github.com/pawamoy/copier-pdm/blob/main/project/scripts/setup.sh) you are using `__pypackages__` by default, but the [PEP582](https://peps.python.org/pep-0582/) was rejected. How is this intended?

That's a good question. I started using PEP 582's `__pypackages__` when it was not yet rejected, and came to like it a lot, so I'm still using `__pypackages__` today (though sometimes I wish I could just `pip install` stuff without having to also pass the `-t` option). That's why the setup script of my template uses `__pypackages__` by default. But since PEP 582 was rejected, I wouldn't say it is recommended practice: `__pypackages__` should only be used if you know what you're doing.

> The 3.10 and 3.12 versions/interpreters are skipped. (having set 3.11 to global with pyenv)
I would have expected to have created three venvs for different pyenv versions.

If neither 3.10 or 3.12 are listed in pyenv's global setting, then these two versions are actually not exposed in your PATH, and PDM can't find them. Therefore they are skipped when running PDM Multirun. My recommendation here is to add all your Python versions installed through pyenv to its global setting: `pyenv global 3.12.x 3.11.x 3.10.x 3.9.x 3.8.x` (replacing each `x` with the right number of course). This way all their shims are in the PATH, and you can call `python3.12`, `python3.11`, `python3.10`, etc., and PDM will find them.

> When creating the venvs first
> Not working when set true, but this is likely an [issue with pdm](https://github.com/pdm-project/pdm/issues/2395).

Yep, I think you always need to create the venvs first, because when you set `venv.in_project` to true, PDM will always create the venv as `.venv` in the project folder. It will not suffix it with the Python version, nor even use the name you provided. The names are just used to map venvs location IIUC, but in this case the location is always the same, whatever the Python version. This is not ideal IMO. What I would prefer, if I were to use venvs again instead of `__pypackages__`, is that PDM automatically suffixes my `.venv` folder with `38`, `39`, etc., and also automatically create it when I install dependencies, so that I don't have to manually create all venvs with `pdm venv create`.

By the way thanks for the one-time donation :heart: :rocket: !

**Reply by laenan8466** [reply]
That helped a lot!

Just a quick thought about this...

One could develop the `setup.py` from [copier-pdm](https://github.com/pawamoy/copier-pdm/blob/main/project/scripts/setup.sh) to:

```shell
#!/usr/bin/env bash
set -e

if ! command -v pdm &>/dev/null; then
    if ! command -v pipx &>/dev/null; then
        python3 -m pip install --user pipx
    fi
    pipx install pdm
fi
if ! pdm self list 2>/dev/null | grep -q pdm-multirun; then
    pdm install --plugins
fi

if [ -n "${PDM_MULTIRUN_VERSIONS}" ]; then
    if [ "${PDM_MULTIRUN_USE_VENVS}" = "1" ]; then
        for version in $PDM_MULTIRUN_VERSIONS; do
            if ! pdm venv --path $version &>/dev/null; then
                pdm venv create --name $version $version
            fi
        done
        preferred_version=$(pdm venv --path $PDM_PREFERRED_VERSION)
        pdm use -f ${preferred_version}
    fi
    pdm multirun -v pdm install -G:all
else
    pdm install -G:all
fi
```

Adding to the `Makefile`:
```make
export PDM_MULTIRUN_USE_VENVS = 1
export PDM_MULTIRUN_VERSIONS ?= 3.8 3.9 3.10 3.11 3.12
export PDM_PREFERRED_VERSION = 3.11
```

To get the desired behavior (multiple venvs in `venv.location`, setting the preferred version as default when using pure `pdm`). The `Makefile` could be compiled from questions during `copier` (e.g. do you want to use `PEP582`?)

Would that be s.th. you would like to see a PR in the copier-pdm repo for?

**Reply by pawamoy** [reply]
Yes! This is definitely something worth adding to the setup script :slightly_smiling_face: Great suggestion!

To be clear, here's what I would accept:

- the modification to the setup script to handle the `PDM_MULTIRUN_USE_VENVS` environment variable
- maybe an additional line that sets `venvs.in_project` to `false`, locally (with `--local`)
- other automation that makes it easier to switch between venvs or pypackages installs (not sure yet what would be the best way)

And here's what I wouldn't accept:

- handling of the `PDM_PREFERRED_VERSION` environment variable: I feel like it's best left to the user. Maybe we can add a rule in the Makefile though, to make it easier, because `pdm use -f $(pdm venv --path 3.11)` is a long command to type :sweat_smile: 
- additional questions in `copier.yml`: I don't want to "fix" the choice between venvs and pypackages into the template answers / project config. I want every individual contributor to be able to choose whether they prefer using venvs or pypackages. This will need proper documentation. I can take care of it in a PR you create :slightly_smiling_face: (if you want to create one! otherwise I'll eventually get to it)

**Reply by laenan8466** [reply]
> (just enabled discussions, and transferred the issue 🙂)

Sorry. 😄 
 
> The idea is to declare and create a few venvs with PDM, then use the same names with PDM Multirun:
> 
> ```shell
> ...
> ```

Was too tired to say that yesterday: This really kicked the things off for my brain. Thanks!

> > The 3.10 and 3.12 versions/interpreters are skipped. (having set 3.11 to global with pyenv)
> > I would have expected to have created three venvs for different pyenv versions.
> 
> If neither 3.10 or 3.12 are listed in pyenv's global setting, then these two versions are actually not exposed in your PATH, and PDM can't find them. Therefore they are skipped when running PDM Multirun. My recommendation here is to add all your Python versions installed through pyenv to its global setting: `pyenv global 3.12.x 3.11.x 3.10.x 3.9.x 3.8.x` (replacing each `x` with the right number of course). This way all their shims are in the PATH, and you can call `python3.12`, `python3.11`, `python3.10`, etc., and PDM will find them.

Not sure. `pyenv global 3.12.x 3.11.x 3.10.x 3.9.x 3.8.x` got me going yesterday, but today switching back to a single global pyenv version is still working... Well - I can accept that and will investigate no further. 😃 

> By the way thanks for the one-time donation ❤️ 🚀 !

I learned so much from your repos only in the last week, increasing my python (environment) skills a lot. This is the least I can do.



> To be clear, here's what I would accept:
> - ...
> - maybe an additional line that sets venvs.in_project to false, locally (with --local)

Guess this is not needed, as pdm seems to [create external venvs for named venvs per default](https://pdm-project.org/latest/usage/venv/#the-location-of-virtualenvs).

> And here's what I wouldn't accept:

Great & clear communication from you 😍 and on second thought far more sensible (🧠 was tired yesterday...). Will see if I can manage a PR today, but feel free to take all ideas to your liking. And thanks for offering taking over the docs.

**Reply by pawamoy** [reply]
> Sorry

No, no, nothing to be sorry about, discussions weren't enabled when you opened the issue, you did nothing wrong.

> Not sure. pyenv global 3.12.x 3.11.x 3.10.x 3.9.x 3.8.x got me going yesterday, but today switching back to a single global pyenv version is still working... Well - I can accept that and will investigate no further. 😃

That's probably because once PDM found each version once, it's able to find it again, even if it's not exposed in PATH anymore. PDM caches interpreters location in `~/.cache/pdm/use_cache.json`.

> I learned so much from your repos only in the last week

The best thing I can read. It throws me back to when I myself learned a lot from someone else's Python template :smile: 

> [create external venvs for named venvs per default](https://pdm-project.org/latest/usage/venv/#the-location-of-virtualenvs).

Nice, didn't know that! Learning from you too hehehe

> Will see if I can manage a PR today

Great, thanks!
